### PR TITLE
Module AzureRM.Bootstrapper: Block user from importing different profiles/modules of different profiles to same session

### PR DESCRIPTION
**Pull request by viananth**
## Description

If user tries to import different profiles or different modules of different profiles to same powershell session using "Use-AzureRmProfile", user will be blocked as it might cause conflicts.

Eg: 
1. Use-AzureRmProfile -profile latest
2. Use-AzureRmProfile -profile '2017-03-09-profile' -module AzureRm.Compute

This needs to be blocked as profile latest's module version is different that the other profile.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1 [hunk 1]
```diff
@@ -949,11 +949,15 @@ function Use-AzureRmProfile
...
 
-      # Block user if they try to import a module to the session where a different version of the same module is already imported
-      if ($null -ne (Get-Module -Name $Module | Where-Object { $_.Version -ne $version} ))
+      # If a different profile's Azure Module was imported, block user
+      $importedModules = Get-Module "Azure*"
+      foreach ($importedModule in $importedModules) 
       {
-        Write-Error "A different version of module $module is already imported in this session. Start a new PowerShell session and retry the operation."
-        return
+        if (($null -ne $ProfileMap.$Profile.$($importedModule.Name)) -and ($ProfileMap.$Profile.$($importedModule.Name) -ne $importedModule.Version))
```
Comment (markcowl) [inline tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1]: Aren't the versions for a particular profile / module an array?  This seems to be treating them as a single value

Comment (viananth) [inline tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1]: That's correct. I'll fix it.

File: tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1 [hunk 2]
```diff
@@ -949,11 +949,23 @@ function Use-AzureRmProfile
...
+      $importedModules = Get-Module "Azure*"
+      foreach ($importedModule in $importedModules) 
       {
-        Write-Error "A different version of module $module is already imported in this session. Start a new PowerShell session and retry the operation."
-        return
+        $versions = $ProfileMap.$Profile.$($importedModule.Name)
+        if ($null -ne $versions)
+        {
+          # We need the latest version in that profile to be imported. If old version was imported, block user and ask to import in a new session
+          $versionEnum = $versions.GetEnumerator()
```
Comment (markcowl) [inline tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1]: can't you just use a for loop here?

Comment (markcowl) [inline tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1]: I'm wondering about the performance of this, how much does this add to the load time of the modules?

Comment (viananth) [inline tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1]: This issue is tracked here: https://github.com/Azure/azure-powershell/issues/3851

**Comment by markcowl** [comment]
@azuresdkci add to whitelist

**Comment by markcowl** [comment]
@azuresdkci retsest this please

**Comment by markcowl** [comment]
@azuresdkci retest this please

**Pull request merged**